PM: add pm cpu shell interface

yongxu-wang15 opened the pull request:
The purpose of introducing this shell interface
1. For systems of the nxp imx9 series with multiple cores, such as imx95(a55+m7+m33), imx943(a55+2*m33+2*m7), its difficult to verify system soc level deep sleep mode which need all cpu into suspend mode in currently residency time suspend work flow, so want to introduce this interface, it will help us to easy to develop low power feathure.


thanks a lot

yongxu-wang15 commented on soc/nxp/imx/imx9/imx95/Kconfig.defconfig.mimx95.m7:
```diff
@@ -55,6 +55,21 @@ if PM
 	default 640
+
+config SHELL
+	default y if !CONFIG_ZTEST
```
sure, have deleted the CONFIG_ZTEST in there, thanks

yongxu-wang15 commented on subsys/pm/pm.c:
```diff
@@ -149,6 +152,13 @@ bool pm_system_suspend(int32_t kernel_ticks)
+		/* Directly enter cpu_idle, no about soc level low power mode */
+		return false;
+	}
+#endif
```
It's indeed a better method. It has been changed.

yongxu-wang15 commented on subsys/pm/pm_cpu_shell.c:
```diff
@@ -0,0 +1,185 @@
+static void pm_suspend_threads_cb(const struct k_thread *thread, void *user_data)
+{
+	char *name;
+	struct thread_event *thread_node;
+
+	thread_node = malloc(sizeof(struct thread_event));
```
as pm_suspend_threads_cb is called by k_thread_foreach, and create thread_event  node for every thread which is saved into global list suspended_threads, and resume them in pm_resume_threads function.
besides, the free is be set in pm_resume_threads flow for every suspended thread node.
thanks

yongxu-wang15 commented on subsys/pm/pm_cpu_shell.c:
```diff
@@ -0,0 +1,185 @@
+	case PM_STATE_SUSPEND_TO_DISK:
+		return "suspend_to_disk";
+	case PM_STATE_SOFT_OFF:
+		return "soft_off";
```
sure, updated

yongxu-wang15 commented on soc/nxp/imx/imx9/imx95/Kconfig.defconfig.mimx95.m7:
```diff
@@ -55,6 +55,19 @@ if PM
 config IDLE_STACK_SIZE
 	default 640
+
+# Enable pm cpu shell and pm status
+config SHELL
+	default y
+
+config PM_CPU_SHELL
+	default y
+
+config STATS
+	default y
+
+config PM_STATS
+	default y
```
I can understand your concern, the SHELL and other PM_CPU_SHEL, its depend on above CONFIG PM define, its optional now, if enabled CONFIG_PM in sample build, the below config will be used.

yongxu-wang15 commented:
Updated:
fixed CI check issue, all check passed now

Thanks

yongxu-wang15 committed:
pm: introduce pm_cpu_shell interface

Introduce a shell for specific cpu core about PM control,
this aligns with existing design flow where the policy decides
the state based on next event timing, locks and latency constraints,
and the system naturally enters low power through the idle thread.

New shell commands under `cpu`:
  -  cpu states			: List supported CPU power states (from devicetree)
  -  cpu available		: For each state/substate, show if available
  -  cpu lock			: Lock specific low power mode
  -  cpu unlock			: Unlock specific low power mode
  -  cpu idle ms		: Sleep the current thread to let idle run PM

Signed-off-by: Yongxu Wang <yongxu.wang@nxp.com>

yongxu-wang15 commented:
Hi @bjarki-andreasen , Thanks for the review and comments, I understand your concern that we can't force power states and suspend its policy transport. 
In this V2 updated, droped unreasonable forced way by 'cpu lps states' and thread suspend logic. Instead, I turn this into add some pm policy management in this shell:
cpu states: list detail information about supported states
cpu idle: this is an simple sleep command, and do nothing expected sleep current shell thread.
cpu available: list available low power mode, the states can be lock/unlock
cpu lock/unlock: via pm_policy_state_lock_get/put to constraint pm states, by this way, it can let core into lower level states with longer time, such as cpu lock standby 0, then cpu idle 1000ms, we can find core stay in STOP rather than SUSPEND mode in imx95 m7 test.

About the shell transport layer, I have removed forced command, and the cpu idle will follow the policy priciple, I think this, for instance, UART should be handled during the device management stage. As long as the uart peripheral is processed before and after suspend, even at the lowest power consumption, it should not affect the actual shell's functionality (otherwise, even if the shell is not enabled, simple output may still have issues).

Please help to point out mistakes if time free

Thanks a lot

ceolin commented on soc/nxp/imx/imx9/imx95/m7/soc.c:
```diff
@@ -175,6 +177,10 @@ void pm_state_exit_post_ops(enum pm_state state, uint8_t substate_id)
+
+#ifdef CONFIG_PM_CPU_SHELL
+	NVIC_EnableIRQ(LPUART3_IRQn);
+#endif
```
I don't like it very much. This special handling here looks some sort of architectural violation. This function shouldn't need to know about it, specially because it can get in here without a shell action.

ceolin commented on subsys/pm/pm.c:
```diff
@@ -149,6 +149,11 @@ bool pm_system_suspend(int32_t kernel_ticks)
+		/* Directly enter cpu_idle, no about soc level low power mode */
+		return false;
+	}
+
```
This does not look correct ... This is just preventing the subsystem to use a power state. You can achieve the same behavior in the right way and without adding any code just getting a lock for all power states.

ceolin commented on subsys/pm/pm_cpu_shell.c:
```diff
@@ -0,0 +1,222 @@
+	default:
+		return "UNKNOWN";
+	}
+}
```
`pm_state` should be reserved to the pm subsys. Actually this function could be in `pm/state.h`

yongxu-wang15 committed:
pm: states: add enum-string conversion helpers

Add two helper functions to convert power management states between
enum and string:

- pm_state_to_string()
- pm_state_from_string()

Signed-off-by: Yongxu Wang <yongxu.wang@nxp.com>

yongxu-wang15 commented:
Updated from @ceolin comments:
1. deleted uart nvic disabled and enable logic in there
2. remove pm_state_to_string and pm_state_from_string int to state.h
3. lock all soc low power mode in SYS_INT, and use cpu idle ms state sub to unlock and enter its low power mode

Thanks

bjarki-andreasen commented on include/zephyr/pm/state.h:
```diff
@@ -407,6 +428,21 @@ static inline const struct pm_state_info *pm_state_get(uint8_t cpu,
 	return NULL;
 }
 
+static inline const char *pm_state_to_str(enum pm_state state)
+{
+	ARG_UNUSED(state);
+
+	return NULL;
+}
+
+static inline int pm_state_from_str(const char *name, enum pm_state *out)
+{
+	ARG_UNUSED(name);
+	ARG_UNUSED(out);
+
+	return 0;
+}
```
These two helpers don't need mocks IMO, they should only be used if PM is used in the first place, so just leave them out of CONFIG_PM is not defined

bjarki-andreasen commented on subsys/pm/pm_cpu_shell.c:
```diff
@@ -0,0 +1,173 @@
+
+
+	/* check if time is available for min residency */
+	if (ms * 1000 < cpu_state->min_residency_us) {
```
A limit on ms needs to be added, ms * 1000 could overflow, though not likely since I doubt anyone wants the shell to idle for months, but the limit also prevents accidentally entering a huge value, killing the shell for hours for example

yongxu-wang15 replied:
add config option to select the max idle time by cpu idle ms, set time as 1hour as default

bjarki-andreasen commented on subsys/pm/pm_cpu_shell.c:
```diff
@@ -0,0 +1,173 @@
+
+	shell_print(sh, "Supported Low Power States:");
+
+	for (state_info = residency_info;
```
Use `ARRAY_FOR_EACH_PTR() {}`

yongxu-wang15 replied:
sure, updated

bjarki-andreasen commented on subsys/pm/pm_cpu_shell.c:
```diff
@@ -0,0 +1,173 @@
+		return -EINVAL;
+	}
+
+	pm_policy_state_lock_put(st, sub);
+
+	k_msleep(ms);
+
+	pm_policy_state_lock_get(st, sub);
```
This can't guarantee entering the state/substate set, unless every other sleep state has been locked, and nothing else is holding this state/substate lock.

The shell really only needs the ability to lock/unlock sleep states, and list which states are and are not locked. Once the state locks are configured, the system will idle, the k_msleep() here should be redundant

bjarki-andreasen reviewed (commented):
Definitely moving in the right direction!

yongxu-wang15 commented:
Updated from @bjarki-andreasen comments
1. removed unused function declaration
2. add cmd_cpu_lock and lock interface now, and cpu idle will just to check and sleep times for currently shell thread
3. use ARRAY_FOR_EACH_PTR instead of for

Thanks

bjarki-andreasen commented on subsys/pm/pm_cpu_shell.c:
```diff
@@ -0,0 +1,204 @@
+		return err;
+	}
+
+	pm_policy_state_lock_get(st, sub);
```
this is a reference counted API, so the shell should make sure the user can not lock the same state multiple times.

pdgendt reviewed (commented):
Thanks, looking good, mostly some nits

ceolin commented on subsys/pm/pm_cpu_shell.c:
```diff
@@ -0,0 +1,164 @@
+		return err;
+	}
+
+	k_msleep(ms);
```
This won't put the cpu to sleep, just the shell thread, I see that you put a comment in the command about it so it should be ok.
